lib/Dialect/XLLVM/XLLVMOps.cpp

```diff
@@ -0,0 +1,68 @@
+//===---- XLLVMOps.cpp - XLLVM Dialect Operations ---------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc.
+//
+//===----------------------------------------------------------------------===//
+// External LLVM (XLLVM) Dialect implementation.
+//===----------------------------------------------------------------------===//
+
+#include "aie/Dialect/XLLVM/XLLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Target/LLVMIR/ModuleTranslation.h"
+#include "mlir/Transforms/FoldUtils.h"
+
+using namespace mlir;
+using namespace xilinx;
+using namespace xilinx::xllvm;
+
+#include "aie/Dialect/XLLVM/IR/XLLVMDialect.cpp.inc"
+
+//===----------------------------------------------------------------------===//
+// XLLVMDialect
+//===----------------------------------------------------------------------===//
+
+void xllvm::XLLVMDialect::initialize() {
+  addOperations<
+#define GET_OP_LIST
+#include "aie/Dialect/XLLVM/IR/XLLVMOps.cpp.inc"
+      >();
+}
+
+namespace xilinx::xllvm {
+
+static llvm::Function *
+getNamedIntrinsicDeclaration(llvm::Module *M, llvm::StringRef fullName,
+                             llvm::Type *resTy, ArrayRef<llvm::Type *> argsTy) {
+  auto *FT = llvm::FunctionType::get(resTy, argsTy, /*isVarArg=*/false);
+  return cast<llvm::Function>(M->getOrInsertFunction(fullName, FT).getCallee());
+}
+
+llvm::CallInst *createExternalLLVMIntrinsicCall(
+    llvm::IRBuilderBase &builder, LLVM::ModuleTranslation &moduleTranslation,
+    Operation *intrOp, llvm::StringRef intrinsicName) {
+  // We support 0 or 1 results
+  assert(intrOp->getNumResults() <= 1 &&
+         "external multi-result intrinsics not supported");
+  llvm::Type *resTy = nullptr;
+  if (intrOp->getNumResults())
+    resTy = moduleTranslation.convertType(*(intrOp->getResultTypes().begin()));
+  auto operands = moduleTranslation.lookupValues(intrOp->getOperands());
```

Comment:
Interesting - I didn't know you could extend llvm translation like this. Cool.

Reply:
The "magic" is happening [during registration of the -mlir-to-llvmir pass](https://github.com/Xilinx/mlir-aie/blob/ce3d8d9215a4c24969363d869f722316620f060e/tools/aie-translate/aie-translate.cpp#L25-L44). A bit hackish, but it's contained and gets the job done.